### ISSUE TITLE
Desfire adpu processing for real iso-8716

### ISSUE THREAD

**Issue by lvandenb** [emsec/ChameleonMini#313]
module MifareDesfireProcess
will not work for real iso - 8716 apdu like
```
00  A4  04  00  0b     a0 00 00 03 97 43 49 44 5f 01 00 9e 32
CLA INS P1  P2  LE
```
right now, it is only testing "wrapped desfire"  or "native". and it fails or the first command...
```
if(ByteCount >= 8 && DesfireCLA(Buffer[0]) && Buffer[2] == 0x00 && 
       Buffer[3] == 0x00 && Buffer[4] == ByteCount - 8) { 
```
in this case buffer[2] == 0x04, and some iso-8716 do not have a  LE field ...

And it seems simple to do 
if Buffer[0]==0x90 then wrapped for sure, else if Buffer[0]<0x0a then most likely 8716  else most likely Desfire Native.
for native check the first byte as a valid command, otherwise the second..

most of the time, communication starts with file select , like "select the desfire app" (this is optional but recommended)
00 A4 04 00 07 D2 76 00 00 85 01 00

on windows, it will always try 00 a4  04  00 0b a0 00 00 03 97 43 49 44 5f 01 00 9e 32, first, if smartcard services are enabled.
( so windows users using the pc/sc api, will always fail trying native desfire when the smartcard service is active)

for the global communication there should be a flag, like "protocolFraming" = unknown /  native / 8716 
after the card  becomes  "state active" , the first command determines the protocol framing. until reset..

**Comment by maxieds** [comment]
@lvandenb 
I will put fixing this on my TODO list. Right now I am still getting up to speed on the first issue #302. It can probably wait for a second pull request once the first is done?

**Comment by maxieds** [comment]
@lvandenb 
>CODEC RX DATA: 0A 01 00 A4 04 00 0B A0 00 00 03 97 43 49 44 5F 01 00 9E 32

Do you know what the prepended two bytes correspond to? How should this get preprocessed?

**Comment by lvandenb** [comment]
these are 14443 I-blocks. The last two bytes will be 14443 crc.

**Comment by maxieds** [comment]
Are the two prologue bytes something that can consistently be identified, or are they something that the reader is free to arbitrarily append to the ISO7816 command format? 

Also, thank you for all the help these last couple of weeks! I am feeling really good about getting the pull request in today (by the end of my marathon weekend) to add complete reader support. All my previous testing was with wrapped and native commands...

**Comment by maxieds** [comment]
@lvandenb 
The [local implementation](https://github.com/emsec/ChameleonMini/compare/master...maxieds:DESFire-NewYears2022-PreliminaryMods) needs some testing. I do not believe either one of my readers is sending out the example command you gave above. The changes made in the last commit DID prevent ``pcsc_scan -v`` from listing that the card is unresponsive. Now that command correctly says that a DESFire card was inserted. This is progress.

**Comment by lvandenb** [comment]
I thought most processing was already done via MifareDesfireProcess()
this keeps the prologue, and adds it back afterwards.

but only wrapped desfire ( class starts with 0x90, or native Desfire is processed now.
(and even iso8716 wrapped is first unwrapped to look like native desfire command.. )

anyway, the first command determines Native or 8716 (including wrapped native)

for the implementation I can make existing send and receive apdu's. 
( with the random number fixed at sender and receiver, it is easier to test.  
I guess the tools on https://pcsclite.apdu.fr/ let you end and receive the apdu's on linux.

**Comment by maxieds** [comment]
I have had my head in the clouds since Thursday on like four hours of sleep working on mathematical writing style to eloquently describe the results in an analytic number theory paper I am responsible for this year. On top of that, it's snowing hard for the usually light hoodie as winter weather in Atlanta. Let me have another cup of coffee and come back to writing code in an hour. 

Perhaps we can aim to get this pull request started for the holiday (in the US) tomorrow? When I file the PR, I will make sure to note at the top that it is collaborative with @lvandenb and @colinoflynn.

**Comment by maxieds** [comment]
>I guess the tools on https://pcsclite.apdu.fr/ let you end and receive the apdu's on linux.

I found the ``scriptor`` [utility docs here](https://ludovicrousseau.blogspot.com/2014/03/level-1-smart-card-support-on-gnulinux.html) so that I do not have to do this in low-level C.

**Comment by maxieds** [comment]
@lvandenb @colinoflynn
Can you help me with testing the code from the latest commit to [my local branch](https://github.com/emsec/ChameleonMini/compare/master...maxieds:DESFire-NewYears2022-PreliminaryMods)? 

The ``scriptor`` command reports an invalid SW byte when transferring the command from above, but the output ``0x6a 0x82`` is accurately returned and prefixed with the prologue bytes.

**Comment by lvandenb** [comment]
I'm still waiting for my Proxmark. 
But below is what happens on windows with  Omnikey tool, and a real Desfire EV1.

**this is sniffed from usb, so the 14443a-4 framing is handled by the reader.**

the full frame there shoud be 2 leading bytes (I-Block), and 2 trailing crc bytes with the answer.
like 02 01 6a 82 crc1 crc2 ?


OUT : 00 a4 04 00 0b a0 00 00 03 97 43 49 44 5f 01 00  
  (ISO File Select P1: 04  (Select by DF Name) , P2: 0 (FCI stored in the file with ID 31) , Data a0000003974349445f0100)
  (looks for Microsoft PNP AID)

IN  : 6a 82  
  file not found


OUT : 00 ca 7f 68 00  
  (Get Data , BER-TLV tag (2 bytes) in P1-P2)

IN  : 6d 00  
  (Instruction not supported)


OUT : 00 a4 04 00 09 a0 00 00 03 08 00 00 10 00  
 (ISO File Select P1: 0 Select by DF Name P2: 0 FCI stored in the file with ID 31 :Data a00000030800001000 )
 ( looks for Personal Identity Verification (PIV) AID) 
IN  : 6a 82 
  ( file not found)

OUT : 00 a4 04 00 09 a0 00 00 03 97 42 54 46 59  
 (ISO File Select P1: 0 Select by DF Name , P2: 0 FCI stored in the file with ID 31 , Data a00000039742544659)
 (looks for Microsoft IDMP AID )
IN  : 6a 82  
  (file not found)


remark: well known AID list in https://www.eftlab.com/knowledge-base/211-emv-aid-rid-pix/

**Comment by maxieds** [comment]
@lvandenb 
Do you think the code is ready to file the pull request? It definitely resolves #302.

**Comment by lvandenb** [comment]
Last night it still did not answer the iso file select correctly
it was 4 bytes 02 01 6a 82  , instead of the needed 02 01 6a 82 crc1 crc2 
Also the log seemed to be overwritten by 6a 82

**Comment by lvandenb** [comment]
I just check the latest build

the iso file select answers  twice , but both are incomplete
1:  6a 82 94 21  :  missing prologue
2:  0a 01 6a 82   :  so missing crc

maybe you can find this bug it faster



Op wo 19 jan. 2022 om 22:38 schreef Maxie D. Schmidt <
***@***.***>:

**Comment by maxieds** [comment]
@lvandenb 
I believe it is fixed now with the latest commit. Please verify (frame 208):

<img src="https://user-images.githubusercontent.com/22165688/150414291-281eaa9a-1e83-4630-bb6f-7212296bac3c.png" width="300" />

**Comment by lvandenb** [comment]
@Maxie,
ok, almost.
but I really think the crc of this answer is calulated over  6A 82
I guess the answer should be 0A 01 6A 82 4D EF   instead of 0A 01 6A 82 9F
21 ?

Op do 20 jan. 2022 om 21:10 schreef Maxie D. Schmidt <
***@***.***>:

**Comment by maxieds** [comment]
@lvandenb 
I cannot verify the precise results you are getting because the ``scriptor`` utility prepends different prologue bytes. The output of running ``scriptor`` on the same command inner data does show an improvement and output the recognized SW error bytes as *File Not Found*. I assume this will now work on your setup as well. Please test with the latest commit to my branch.

You were right. I computed the CRC bytes on just the data part of the buffer before prepending the prologue back to it prior to returning. This was actually a point of confusion for me initially. I thought that it might want a consistency check on just the data for application processing. 

At any rate, cheers! It appears to be working. 😺

**Comment by lvandenb** [comment]
Ok this is looking good.

right now only
[60804] CODEC RX DATA: 0B 01 00 CA 7F 68 00 73 73 (read data )
[60804] CODEC TX DATA: 0B 01 **6A 81** 6D C1  (Function not supported)

but a real desfire ev1 card responds 
0B 01 **6D 00** E3 19  (6D00 : Instruction code not supported or invalid)

**Comment by maxieds** [comment]
@lvandenb 
What do you think about creating the pull request to fix the start of this issue and #302 right now? Maybe we should start a new thread about more subtle bugs in the DESFire implementation like the one you just pointed out.

**Comment by lvandenb** [comment]
@maxie 
yes, we can close these now.
its another level now.

**Comment by maxieds** [comment]
@lvandenb 
Please open a new issue for incremental debugging of DESFire emulation and paste the previous message with the data showing an incorrect response there. 

Thanks again for all the help debugging and testing.

**Comment by maxieds** [comment]
@lvandenb 
I have been looking at the last command discrepancy you noted. The LE field in the command is zero, correct? That might cause an error in [this code](https://github.com/emsec/ChameleonMini/blob/master/Firmware/Chameleon-Mini/Application/MifareDESFire.c#L185). I have to look at it more carefully. 

Are you planning to do more DESFire testing with the Chameleon once your Proxmark arrives?

**Comment by fptrs** [comment]
resolved by #314

**Comment by lvandenb** [comment]
Yes, I'll do further testing 
The proxmark is on its way now. 
Expected somewhere next week

**Comment by david-oswald** [comment]
Hi, I just tried the most recently merged PR with a Proxmark (using the current GIT head of the RRG proxmark), first trying to get the application list in different communication modes:

```
[usb] pm3 --> hf mfdes getaids -a -v -m plain
[=] Key num: 0 Key algo: des Key[8]: 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: niso Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] Anticollision ok
[=] Auth: cmd: 0x1a keynum: 0x00
[+] >>>> 90 1A 00 00 01 00 00
[!!] 🚨 APDU: No APDU response
[!!] 🚨 Desfire authenticate error. Result: [1] Sending auth command failed
[+] Setting ISODEP -> inactive
[usb] pm3 --> hf mfdes getaids -a -v -m plain -c native
[=] Key num: 0 Key algo: des Key[8]: 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] Anticollision ok
[=] Auth: cmd: 0x1a keynum: 0x00
[+] raw>> 1A 00
[!!] 🚨 Desfire authenticate error. Result: [1] Sending auth command failed
[+] Setting ISODEP -> inactive
```

It seems to fail on the authentication step. I got a log from the Chameleon, this is what Chamlog has to say:

```
38784 ms <+38784 ms>:UNKNOWN                      (2   bytes) [1159                ]
38856 ms <   +72 ms>:CODEC RX                     (1   bytes) [52                  ]
38856 ms <    +0 ms>:CODEC TX                     (2   bytes) [4403                ]
38857 ms <    +1 ms>:CODEC RX                     (2   bytes) [9320                ]
38857 ms <    +0 ms>:CODEC TX                     (5   bytes) [880800e262          ]
38858 ms <    +1 ms>:CODEC RX                     (9   bytes) [9370880800e2620c76  ]
38858 ms <    +0 ms>:CODEC TX                     (3   bytes) [24d836              ]
38859 ms <    +1 ms>:CODEC RX                     (2   bytes) [9520                ]
38859 ms <    +0 ms>:CODEC TX                     (5   bytes) [0a6a7a968c          ]
38861 ms <    +2 ms>:CODEC RX                     (9   bytes) [95700a6a7a968cf878  ]
38861 ms <    +0 ms>:CODEC TX                     (3   bytes) [20fc70              ]
38861 ms <    +0 ms>:CODEC RX                     (4   bytes) [e0803173            ]
38861 ms <    +0 ms>:CODEC TX                     (8   bytes) [06750081028066fd    ]
38886 ms <   +25 ms>:CODEC RX                     (10  bytes) [02901a0000010000e35c]
44378 ms < +5492 ms>:CODEC RX                     (1   bytes) [52                  ]
44378 ms <    +0 ms>:CODEC TX                     (2   bytes) [4403                ]
44379 ms <    +1 ms>:CODEC RX                     (2   bytes) [9320                ]
44379 ms <    +0 ms>:CODEC TX                     (5   bytes) [880800e262          ]
44380 ms <    +1 ms>:CODEC RX                     (9   bytes) [9370880800e2620c76  ]
44380 ms <    +0 ms>:CODEC TX                     (3   bytes) [24d836              ]
44381 ms <    +1 ms>:CODEC RX                     (2   bytes) [9520                ]
44381 ms <    +0 ms>:CODEC TX                     (5   bytes) [0a6a7a968c          ]
44383 ms <    +2 ms>:CODEC RX                     (9   bytes) [95700a6a7a968cf878  ]
44383 ms <    +0 ms>:CODEC TX                     (3   bytes) [20fc70              ]
44383 ms <    +0 ms>:CODEC RX                     (4   bytes) [e0803173            ]
44383 ms <    +0 ms>:CODEC TX                     (8   bytes) [06750081028066fd    ]
44408 ms <   +25 ms>:CODEC RX                     (6   bytes) [0b001a00f4fe        ]
```

So it seems the codec does not reply to the auth attempt at all in both communication modes, any quick ideas what I might be doing wrong?

**Comment by maxieds** [comment]
@david-oswald 
You might have used a case I didn't test when writing the first part of the implementation in #287. The LibNFC [test code here](https://github.com/emsec/ChameleonMini/blob/master/Software/DESFireLibNFCTesting/LocalInclude/DesfireUtils.h#L19) uses the AuthAES command ``0xAA``. That should definitely be working. Does your query go through if you replace ``0x1A`` with ``0xAA``? The DESFire source to handle the command ``0x1A`` (as you used above) is [here](https://github.com/emsec/ChameleonMini/blob/master/Firmware/Chameleon-Mini/Application/DESFire/DESFireInstructions.c#L1669). If I have some time later today, I will try to test with the AuthISO command to see if I can debug it. It will be helpful to know if your Proxmark reports success on ``0xAA`` though.

**Comment by lvandenb** [comment]
I guess it can be tested with a prologue of 2 or 1 bytes, and the wrapped or native version
prologue 1 byte , wrapped :  02 90 0a 00 crc crc
prologue 2 bytes , wrapped :  A0 00 90 0a 00 crc crc
prologue 1 byte , native :  02 0a 00 crc crc
prologue 2 bytes , native:  A0 00 0a crc crc



anyway the 0x0A native authentication would need the first to work, 
because an empty card always authenticates with 0x0A, with desfire key 16 times 00
(which has to be considered as single des, because 2nd 8 bytes equals the first 8 bytes)

to be able to test the Authenticate AES  0xAA (on the PICC), we would need the Authenticate command  0X0A key 0 , followed  by  ChangeKey command (0x54) , key 0x80 (to AES)

**Comment by maxieds** [comment]
Maybe a second or third pair of eyes can help me figure out what's going wrong with [the code in my new branch](https://github.com/maxieds/ChameleonMini/tree/DESFire-AuthISO-Patch). Perhaps OpenSSL on the host side is producing weird values?

**Comment by maxieds** [comment]
@david-oswald 
I believe that the ``AuthISO(0x1A)`` command is now working with the most recent commits to [my local branch](https://github.com/maxieds/ChameleonMini/tree/DESFire-AuthISO-Patch). Can you please test to confirm it works with the Proxmark?

I wrote LibNFC test code that runs on my Linux host (see [Software/ directory here](https://github.com/maxieds/ChameleonMini/tree/DESFire-AuthISO-Patch/Software/DESFireLibNFCTesting)). It's kind of strangely behaving running off my CMLD application connected to my phone. I have better luck getting the command to go through with the Chameleon on battery. The strange thing is that almost always, if I run an ``AuthAES(0xAA)`` and that succeeds, then immediately after the ``AuthISO(0x1A)`` will go through:
```bash
$ cd Software/DESFireLibNFCTesting
$ make
$ ./Bin/TestAuthenticateAES128.exe
## Output ...
$ ./Bin/TestAuthenticateISO.exe
## More output ...
```
I wonder if there is a bug in the code where it wants the AES auth procedure to do what @lvandenb noted that the ``AuthLegacy(0x0A)`` should be doing? Otherwise, the implementation for 3DES is less efficient than the AES HW enhanced code? I'm not sure what it's doing there...

I am going to try to get the legacy auth code tested and working this morning. If some testers can confirm that it all goes well with them as well, I think we can file another PR over the next couple of days to enhance the DESFire emulation support. 😸

**Comment by maxieds** [comment]
If it's easier, once I get the legacy authentication command working, I can upload a binary of the new firmware build so testers do not need to mess with cloning and building the code in my local branch. Just in case this gets more folks motivated, here's how to do that yourselves:
```bash
$ git clone https://github.com/maxieds/ChameleonMini.git
$ cd ChameleonMini
$ git checkout DESFire-AuthISO-Patch
$ cd Firmware/Chameleon-Mini
$ make desfire
$ sudo avrdude -c flip2 -p ATXMega128A4U -B 60 -P usb -U application:w:Chameleon-Mini.hex:i -U eeprom:w:Chameleon-Mini.eep:i
```

**Comment by maxieds** [comment]
Anyone tried this out yet? Should I just go ahead and file the PR? The code in my branch definitely fixes some problems with two of the auth command variants and adds some functionality into the Software / LibNFC-based test suite.

**Comment by david-oswald** [comment]
I tried with various variations on the proxmark cmdline, but it always seems to fail. I have not much time to debug this today/tomorrow unfortunately...

**Comment by maxieds** [comment]
@david-oswald 
What sort of exchanges does the Proxmark expect? Does it need to be MAC'ed and encrypted for all commands handled after the authentication? The MAC'ed exchange is something on a TODO list to add to the source. Right now all the emulation support can handle is cleartext exchange. It was very spec dependent so I could use some pointers about what it needs to be doing. IIRC, it is not clearly documented, at least publicly, which exchange protocol, crypto algorithm, etc. to invoke when presented with ISO / versus native wrapped / other handshakes. The current implementation defaults to non-MAC'ed cleartext transfers only.

**Comment by david-oswald** [comment]
@maxieds The Proxmark DESFire command set is documented here: https://github.com/RfidResearchGroup/proxmark3/blob/master/doc/desfire.md

I think the Proxmark reader code is also a good "documentation in code" for the future.

I tried many variations of channel, iso/niso/native, crypto, etc. One example of just a plain auth:

```
[usb] pm3 --> hf mfdes auth -n 0 -t des -k 0000000000000000 -f none -v -c native -a
[=] Key num: 0 Key algo: des Key[8]: 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] AID 000000 is selected
[=] Auth: cmd: 0x1a keynum: 0x00
[+] raw>> 1A 00
[!!] 🚨 Desfire authenticate error. Result: [1] Sending auth command failed
[+] Setting ISODEP -> inactive
[-] ⛔ Select or authentication AID 000000 failed. Result [1] Sending auth command failed
[usb] pm3 --> hf mfdes auth -n 0 -t des -k 0000000000000000 -f none -v -c niso -a
[=] Key num: 0 Key algo: des Key[8]: 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: niso Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] AID 000000 is selected
[=] Auth: cmd: 0x1a keynum: 0x00
[+] >>>> 90 1A 00 00 01 00 00
[!!] 🚨 APDU: No APDU response
[!!] 🚨 Desfire authenticate error. Result: [1] Sending auth command failed
```

**Comment by maxieds** [comment]
@david-oswald 
I really need to get my hands on one of the pricier Proxmarks to do testing. Can you quickly try to verify if the AES authentication works sending command ``0xAA``? If that goes through, then it's a bug that should be easier for me to track down.

**Comment by david-oswald** [comment]
Not really fixing it - I always also try "native ISO" as seen above (the `-c` param) and that doesn't work either... Is there any logging I could turn on to help you debug further?

```
[usb] pm3 --> hf mfdes auth -n 0 -t des -k 0000000000000000 -f none -v -c native -a
[=] Key num: 0 Key algo: des Key[8]: 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] AID 000000 is selected
[=] Auth: cmd: 0x1a keynum: 0x00
[+] raw>> 1A 00
[!!] 🚨 Desfire authenticate error. Result: [1] Sending auth command failed
[+] Setting ISODEP -> inactive
[-] ⛔ Select or authentication AID 000000 failed. Result [1] Sending auth command failed

[usb] pm3 --> hf mfdes auth -n 0 -t des -k 0000000000000000 -f none -v -c niso -a
[=] Key num: 0 Key algo: des Key[8]: 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: niso Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] AID 000000 is selected
[=] Auth: cmd: 0x1a keynum: 0x00
[+] >>>> 90 1A 00 00 01 00 00
[!!] 🚨 APDU: No APDU response
[!!] 🚨 Desfire authenticate error. Result: [1] Sending auth command failed
[+] Setting ISODEP -> inactive
[-] ⛔ Select or authentication AID 000000 failed. Result [1] Sending auth command failed
```

**Comment by maxieds** [comment]
First of all, enable these settings in the ``Makefile`` when you compile the firmware:
```bash
SETTINGS  += -DDESFIRE_DEFAULT_TESTING_MODE=1
SETTINGS  += -DDESFIRE_CRYPTO1_SAVE_SPACE
```
I personally get a lot out of using the [CMLD app](https://github.com/maxieds/ChameleonMiniLiveDebugger) I wrote with the (Droid) phone connected to the Chameleon over a wired serial USB connection. If you turn on the LIVE logging feature, then you should get quite a bit of formatted output while the auth command is in transit. You can install CMLD via the Google Play Store as usual for free. Here is a screenshot of where to turn on the LIVE logging from the phone navigation and what I get running the ``0x1A`` auth with LibNFC on the host with logging to the phone:

<img src="https://user-images.githubusercontent.com/22165688/151842858-d3acfdc3-5636-470a-a76c-eb9f39433b2d.png" width="250" /><img src="https://user-images.githubusercontent.com/22165688/151846436-6bb5ed75-ff7f-4ab0-98ba-1fc41219e6cd.png" width="250" /><img src="https://user-images.githubusercontent.com/22165688/151846433-abf9ae5f-c0d8-4162-8e64-701b42faccd5.png" width="250" />

I still wonder if you can get it to go through by using AES auth ``0xAA`` on the Proxmark? I have luck with that running in front of the other Auth commands with the LibNFC test code. If that for some reason does work, then the list of possibilities for bugs to check in the code is substantially narrowed... Please also remember to run ``git pull`` for my branch and rebuild. I just pushed something else that I have been working on.

**Comment by maxieds** [comment]
I have another quick implementation question. Are the rndA and rndB buffers used for authentication with AES and 3K3DES supposed to be 16 bytes? Right now the implementation I have uses 8 byte buffers for rndA and rndB. If they should be 16 bytes instead, I need to modify my working code.

**Comment by maxieds** [comment]
The [Proxmark3 source code](https://github.com/RfidResearchGroup/proxmark3/blob/2269ad69147bb947130bbdd86ed4a208074e2616/client/src/mifare/desfirecore.c#L1101) indicates that we need to use 16 bytes for rndA and rndB. I will change the source code.

**Comment by lvandenb** [comment]
8 byte for DES or 2K3DES , 16 byte 3k3des and AES.  so it depends on the key lenght also...

**Comment by maxieds** [comment]
@david-oswald 
Do you mind checking the ISO authentication again with the [latest commit to my local branch](https://github.com/maxieds/ChameleonMini/tree/DESFire-AuthISO-Patch)? I put a lot of effort in this week debugging...

**Comment by lvandenb** [comment]
van this issue be reopened ? Maybe move the authentication to another issue
(0x1a should give an error on initial Desfire EV1 , so I would need the authenticate native and changekey first)

=> there are still some small issues with 7816 responses
6a 82 should be for file not found, the 6a 81 for unknown instruction, but then it starts responding both ?
![image](https://user-images.githubusercontent.com/3081470/152674467-09b7a690-99e5-44fc-8601-ae25142a42b8.png)

**Comment by david-oswald** [comment]
> @david-oswald Do you mind checking the ISO authentication again with the [latest commit to my local branch](https://github.com/maxieds/ChameleonMini/tree/DESFire-AuthISO-Patch)? I put a lot of effort in this week debugging...

@maxieds I did try that, but something in your recents commits to that branch seems to have introduced a new problem - on doing an auth or even just aid list, the Chameleon completely crashes and does not respond to anything further. I also noticed that the output of cmdline commands like `help` appears truncated.

**Comment by maxieds** [comment]
@david-oswald 
I noticed the crashes with the legacy authenticate. Not sure if that is what you were using. Can you try again with the modifications in the latest commit. I am getting my Proxmark RDV4 finally shipped out to me in a week or two. That will let me do this sort of bothersome testing / back and forth for myself when it arrives. 

@lvandenb 
I will try to get to your issues with the ISO7816 return codes later on tonight.

**Comment by maxieds** [comment]
@lvandenb 
Does the legacy authenticate need to be done with DES, or can it be set to default to 3DES? The current implementation uses 3DES.

**Comment by lvandenb** [comment]
It is 2k3des. But when first 8 bytes equal second 8, it is Des.
Also native 2k3des is done differently than iso3k3des.
Native is always EEE on card, even for decryption.


Op za 12 feb. 2022 04:18 schreef Maxie D. Schmidt ***@***.***

**Comment by maxieds** [comment]
@lvandenb 
The latest commit to my local branch appears to pass all of the LibNFC-based tests I put it through with the ACR reader. It should be ready for you or @david-oswald to test with the Proxmark. Again, sorry for having to rely so much on you for doing testing I should be able to do. My Proxmark is arriving on Tuesday...

The ISO-7816 return codes problems you mentioned above are on my TODO, but not yet urgent, coding projects list.

**Comment by maxieds** [comment]
@lvandenb 
>=> there are still some small issues with 7816 responses
>6a 82 should be for file not found, the 6a 81 for unknown instruction, but then it >starts responding both ?

Are you sure that all of the RDR exchanges are getting accurately logged? It looks like the mix-up in SW bytes returned corresponds to two different transaction types. Check the prologue bytes on the TAG responses. The return codes seem accurate for the prologue bytes associated with the previous commands with those same prologue bytes. IDK why it would be non-deterministic in that way otherwise. Maybe the Chameleon is dropping some of the logs or some of the RDR incoming commands if new ones arrive to process next? We could look into the codec code, but there are other devs that would know those internals better than I do.

Do you mind checking the latest commit to my local branch to see if the ACK/NAK phenomenon on the NXP-based readers is now resolved?

**Comment by lvandenb** [comment]
Sorry, now I noticed that the two switches on the PM3 were left in the
worst position
(125 kHz and worst Q factor).
I'll do new testing with latest Repo.

Op za 12 feb. 2022 14:07 schreef Maxie D. Schmidt ***@***.***

**Comment by maxieds** [comment]
@lvandenb 
Do you mind testing the [latest firmware commit](https://github.com/maxieds/ChameleonMini/tree/DESFire-AuthISO-Patch) with your PM3? The commands I am running are:
```bash
hw dbg -4
prefs set clientdebug --full
data setdebugmode -2
hf mfdes auth -n 0 -t des -k 0000000000000000 -v -c native -a
```
I am not able to test this patch with my PM3 right now. I am going to get the battery pack for my device since it is being very temperamental with my self-powered USB hub. I think the firmware should finally work.

**Comment by david-oswald** [comment]
Hi @maxieds that sounds great! I will have a play with it next week. Also thanks for the PR - @fptrs who normally does the merges will be away for some time, so I'm not sure if he still has time to review, but otherwise I might've a look as it definitely seems to improve the state of the DESFire impl.

**Comment by maxieds** [comment]
@david-oswald 
Any luck testing the newest DESFire firmware mods?

**Comment by colinoflynn** [comment]
@maxieds I didn't realize there was an open PR for desfire, I ported my "mfp" hacks to latest master & caught a few more bugs. I didn't see some of them in your PR, but they may have been 'fixed' elsewhere so wanted to check. In particular:

* Bug fixes only are in 3 commits at https://github.com/colinoflynn/ChameleonMini/commits/desfire-fixes
* I updated my code in https://github.com/colinoflynn/ChameleonMini/commits/mpev1seperateloop but will need to update again - looks like you've got AES CMAC code (I had added a simple version). One other item was the AES accelerator seemed to  go into infinite interrupts, I had turned off the interrupt flag for it (not sure if you ran into this). Interrupts weren't used from what I could see with the AES accelerator anyway (the callback is never set).

Thanks again for this! My latest MFP code is more isolated from the DESFire command processing to make it less intrusive if people want to expand in the future.

**Comment by maxieds** [comment]
@colinoflynn 
I incorporated your bug fixes into the code for this pull request. I am slammed right now with work and things to get done. I am also not sure how to adequately test that the bugfixes work. Do you mind compiling the firmware from the most recent commit to my fork/branch and testing? 😸

**Comment by maxieds** [comment]
@colinoflynn @david-oswald 
Any updates with testing? Any chance this PR can get merged soon? The Spring 2022 semester at GT is over. This was my RA project for the term. It would be nice to see it in production in case anyone here asks. 😺

**Comment by lvandenb** [comment]
Finally got some time, but I have an issue with the Makefile now
git https://github.com/maxieds/ChameleonMini.git
DESFire-AuthISO-Patch


$ make desfire
../LUFA/Build/lufa_build.mk:131: The XMEGA device support is currently EXPERIMENTAL (incomplete and/or non-functional), and is included for preview purposes only.
Makefile:273: *** recipe commences before first target.  Stop.

**Comment by maxieds** [comment]
@lvandenb 
I just tried compiling the sources with ``make desfire`` on MacOS with no problems. The trailing lines after the build completes are
```bash
 [OBJCPY]  : Extracting HEX file data from "Chameleon-Mini.elf"
avr-objcopy -O ihex -R .eeprom -R .fuse -R .lock -R .signature --set-section-flags=.flashdata="alloc,load" Chameleon-Mini.elf Chameleon-Mini.hex
 [OBJCPY]  : Extracting EEP file data from "Chameleon-Mini.elf"
avr-objcopy -O ihex -j .eeprom --set-section-flags=.eeprom="alloc,load" --change-section-lma .eeprom=0 --no-change-warnings Chameleon-Mini.elf Chameleon-Mini.eep || exit 0
 [OBJCPY]  : Extracting BIN file data from "Chameleon-Mini.elf"
avr-objcopy -O binary -R .eeprom -R .fuse -R .lock -R .signature --set-section-flags=.flashdata="alloc,load" Chameleon-Mini.elf Chameleon-Mini.bin

Chameleon-Mini.elf  :
section                       size       addr
.flashdata                 0x10000    0x10000
.spmhelper                    0x14    0x21fe0
.text                       0xe65a        0x0
.data                        0x162   0x802000
.bss                         0xc6d   0x802162
.eeprom                       0x64   0x810000
.comment                      0x26        0x0
.note.gnu.avr.deviceinfo      0x40        0x0
.debug_aranges               0x160        0x0
.debug_info                 0x29f1        0x0
.debug_abbrev               0x2708        0x0
.debug_line                  0x71e        0x0
.debug_str                  0x1b33        0x0
Total                      0x269b1


   text	   data	    bss	    dec	    hex	filename
 0xe65a	  0x1c6	  0xc6d	  62605	   f48d	Chameleon-Mini.elf

AVR Memory Usage
----------------
Device: Unknown

Program:   59324 bytes
(.text + .data + .bootloader)

Data:       3535 bytes
(.data + .bss + .noinit)

EEPROM:      100 bytes
(.eeprom)
```
The warning from the LUFA make script is normal. It has been printed for years. Looking at line ``273`` of the [Makefile](https://github.com/maxieds/ChameleonMini/blob/DESFire-AuthISO-Patch/Firmware/Chameleon-Mini/Makefile#L273) suggests a couple of things. Please fill in more details:
* What platform are you working in?
* What is your terminal and shell? What output do you get running ``which bash``?

A quick fix is to use the binaries I just compiled for testing: [current-build.zip](https://github.com/emsec/ChameleonMini/files/8665156/current-build.zip)

Happy to follow up otherwise.

**Comment by colinoflynn** [comment]
I ran the test by just compiling with `make desfire` and `make desfire-dev` (I tried both). The only configuration I did was setting `python chamtool.py -p COM4 -c mf_desfire`, and I'd check if it was still responding with `python chamtool.py -p COM4 -U` to print the UID in use.

With this configuration, I see:
* Using the command sequence in @maxieds March 30 example, and running that on my pm3 (the full 'hw dbg -4'... etc) **works**. I get the same output showing successful authentication.
* Using just the `hf mfdes auth` command without first doing the debug stuff gives me an error - is this a timing issue or just some invalid thing? It looks like the only difference was the debug stuff in pm3 being enabled, it didn't look like any differences in the actual authentication? Here is the output:
```
[usb] pm3 --> hf mfdes auth -n 0 -t 3tdea -k 000000000000000000000000000000000000000000000000 -v -c native -a
[=] Key num: 0 Key algo: 3tdea Key[24]: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[!!] Desfire AID select error.
[+] Setting ISODEP -> inactive
[-] Select or authentication AID 000000 failed. Result [202] Can't select application by ISO ID.
```

Trying to use `hf mfdes info` causes the ChamelonMini to lock up - it stops responding to my ChamTool requests until I power cycle it. in pm3 I see:
```
[usb] pm3 --> hf mfdes info
[#] pcb_blocknum 0 == 2
[#] [WCMD <--: : 08/08] 02 90 60 00 00 00 14 98
[#] fukked
[#] switch_off
[!] Command unsuccessful
```

It looks like it's locking up on `GET VERSION`:

```
[=] downloading tracelog data from device
[+] Recorded activity (trace len = 178 bytes)
[=] start = start of start frame end = end of frame. src = source of transfer
[=] ISO14443A - all times are in carrier periods (1/13.56MHz)

      Start |        End | Src | Data (! denotes parity error)                                           | CRC | Annotation
------------+------------+-----+-------------------------------------------------------------------------+-----+--------------------
          0 |        992 | Rdr |52                                                                       |     | WUPA
       2100 |       4468 | Tag |44  03                                                                   |     |
       7040 |       9504 | Rdr |93  20                                                                   |     | ANTICOLL
      10548 |      16436 | Tag |88  08  19  ad  34                                                       |     |
      19072 |      29600 | Rdr |93  70  88  08  19  ad  34  9a  9d                                       |  ok | SELECT_UID
      30644 |      34164 | Tag |24  d8  36                                                               |     |
      35584 |      38048 | Rdr |95  20                                                                   |     | ANTICOLL-2
      39092 |      44916 | Tag |3b  4b  ed  57  ca                                                       |     |
      47616 |      58080 | Rdr |95  70  3b  4b  ed  57  ca  b9  c5                                       |  ok | SELECT_UID-2
      59188 |      62772 | Tag |20  fc  70                                                               |     |
      64512 |      69280 | Rdr |e0  80  31  73                                                           |  ok | RATS
      70324 |      77364 | Tag |06  75  00  81  02  80                                                   | !crc|
      94976 |     104352 | Rdr |02  90  60  00  00  00  14  98                                           |  ok | GET VERSION
[usb] pm3 -->
```

Anyway the authentication code seems to be working per the instructions & that is great! And still useful for me to build on this for MFP.

**Comment by maxieds** [comment]
@colinoflynn 
See if the latest commit works for you? An up to date version of the binaries is here: 
[current-build.zip](https://github.com/emsec/ChameleonMini/files/8687030/current-build.zip)

**Comment by colinoflynn** [comment]
Hi @maxieds - this seemed to get further, testing shows this result. It does still lock up (stops responding on `ChamTool`). For some reason I don't get any useful output with `hf mfdes list` (like I did before).

```
[usb] pm3 --> hf mfdes info

[=] ---------------------------------- Tag Information ----------------------------------
[+]               UID: 08 71 98 85 E4 07 E9
[+]      Batch number: 00 00 00 00 00
[+]   Production date: week 00 / 2000

[=] --- Hardware Information
[=]    raw: 63000000000000
[=]      Vendor Id: Pragmatic Printing Ltd. UK
[=]           Type: 0x00
[=]        Subtype: 0x00
[=]        Version: 0.0 ( DESFire MF3ICD40 )
[=]   Storage size: 0x00 ( 1 bytes )
[=]       Protocol: 0x00 ( Unknown )

[=] --- Software Information
[=]    raw: C1A0C900000000
[=]      Vendor Id: no tag-info available
[=]           Type: 0xA0
[=]        Subtype: 0xC9
[=]        Version: 0.0
[=]   Storage size: 0x00 ( 1 bytes )
[=]       Protocol: 0x00 ( Unknown )

[=] --------------------------------- Card capabilities ---------------------------------
[#] error DESFIRESendRaw
[#] error DESFIRESendRaw
[#] error DESFIRESendRaw
[#] Card didn't answer to CL1 select all
[!!] APDU: No APDU response
[#] error DESFIRESendApdu Unknown error
[+] ------------------------------------ PICC level -------------------------------------
[+] Applications count: 0 free memory n/a
[+] PICC level auth commands: auth: NO auth iso: NO auth aes: NO auth ev2: NO auth iso native: NO auth lrp: NO

[=] --- Free memory
[+]    Card doesn't support 'free mem' cmd
```

I'll see if I can get better debug output to show what is giving it problems. BTW I forgot you mentioned missing a proxmark at one point, if can help loan one if helpful let me know.

The authentication side seems to work great, so still major improvements overall. Also the build worked OK now for me too from a clean checkout.

**Comment by maxieds** [comment]
@colinoflynn 
I have some good news to report on progress:
```bash
[usb] pm3 --> hf mfdes info

[=] ---------------------------------- Tag Information ----------------------------------
[+]               UID: 08 61 44 FE B4 2B F9 
[+]      Batch number: 00 00 00 00 00 
[+]   Production date: week 00 / 2000

[=] --- Hardware Information
[=]    raw: EC720000000000
[=]      Vendor Id: no tag-info available
[=]           Type: 0x72
[=]        Subtype: 0x00
[=]        Version: 0.0 ( DESFire MF3ICD40 )
[=]   Storage size: 0x00 ( 1 bytes )
[=]       Protocol: 0x00 ( Unknown )

[=] --- Software Information
[=]    raw: 91C1ACC7000000
[=]      Vendor Id: no tag-info available
[=]           Type: 0xC1
[=]        Subtype: 0xAC
[=]        Version: 199.0
[=]   Storage size: 0x00 ( 1 bytes )
[=]       Protocol: 0x00 ( Unknown )

[=] --------------------------------- Card capabilities ---------------------------------
[#] Halt error
[#] Card didn't answer to CL1 select all
```
I introduced a bug somewhere into the  auth command: 
```hf mfdes auth -n 0 -t 3tdea -k 000000000000000000000000000000000000000000000000 -v -c native -a``` 
I did notice earlier today as you pointed out above (before the latest commit with new fixes) that the auth command is buggy without the debugging options getting set by my script. This behavior has now changed to:
```bash
[usb] pm3 --> script run test-mfdes-auth.cmd 
[+] executing Cmd test-mfdes-auth.cmd
[+] args ''
[usb|script] pm3 --> hf mfdes auth -n 0 -t 3tdea -k 000000000000000000000000000000000000000000000000 -v -c native -a
[=] Key num: 0 Key algo: 3tdea Key[24]: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] AID 000000 is selected
[=] Auth: cmd: 0x1a keynum: 0x00
[+] raw>> 1A 00 
[+] raw<< AF EE 91 30 1E E8 F5 84 D6 C7 85 1D 05 65 13 90 A6 C6 D5 
[#] encRndB: EE 91 30 1E E8 F5 84 D6 
[#] RndB: CA FE BA BE 00 11 22 33 
[#] rotRndB: FE BA BE 00 11 22 33 CA FE BA BE 00 11 22 33 CA 
[#] Both   : 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 FE BA BE 00 11 22 33 CA FE BA BE 00 11 22 33 CA 
[+] raw>> AF 30 EB 55 F3 29 39 04 96 77 88 CE EF 33 A3 C8 7B 18 66 1A F1 62 78 A0 28 53 84 67 98 7C BB DB 03 
[+] raw<< 7E 4B 16 
[!!] 🚨 Command (af) ERROR: 0x7e
[#] error DESFIRESendRaw Length of command string invalid
[!!] 🚨 Desfire authenticate error. Result: [7] Sending auth command failed
[+] Setting ISODEP -> inactive
[-] ⛔ Select or authentication AID 000000 failed. Result [7] Sending auth command failed
[usb|script] pm3 --> script run ./debug.cmd
[+] executing Cmd ./debug.cmd
[+] args ''
[usb|script] pm3 --> hw dbg -4
[usb|script] pm3 --> prefs set clientdebug --full
[=]     client debug........... full
[usb|script] pm3 --> data setdebugmode -2
[=] client debug level... 2 ( verbose debug messages )
[usb|script] pm3 --> hf mfdes auth -n 0 -t 3tdea -k 000000000000000000000000000000000000000000000000 -v -c native -a
[=] Key num: 0 Key algo: 3tdea Key[24]: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[#]   Debug log level......... 4 ( extended )
[#] Card didn't answer to select
[!!] 🚨 Desfire AID select error.
[+] Setting ISODEP -> inactive
[-] ⛔ Select or authentication AID 000000 failed. Result [202] Can't select application by ISO ID.
```
I should have some more time to finish up and get the auth command working tonight or tomorrow :)

Anything else I have missed in all of the excellent feedback you gave above?

**Comment by maxieds** [comment]
@colinoflynn 
Please check with the latest commit to verify. This is what I get with the PM3 after building with ``make desfire-dev``:
```bash
[usb] pm3 --> script run debug.cmd
[+] executing Cmd debug.cmd
[+] args ''
[usb|script] pm3 --> hw dbg -4
[usb|script] pm3 --> prefs set clientdebug --full
[=]     client debug........... full
[usb|script] pm3 --> data setdebugmode -2
[=] client debug level... 2 ( verbose debug messages )

[#]   Debug log level......... 4 ( extended )
[usb] pm3 --> hf mfdes info
[#] pcb_blocknum 0 == 2 
[#] [WCMD <--: : 08/08] 02 90 60 00 00 00 14 98 
[#] pcb_blocknum 1 == 3 
[#] [WCMD <--: : 08/08] 03 90 af 00 00 00 1f 15 
[#] pcb_blocknum 0 == 2 
[#] [WCMD <--: : 08/08] 02 90 af 00 00 00 34 11 

[=] ---------------------------------- Tag Information ----------------------------------
[+]               UID: 08 4F 8A 44 7D AE 83 
[+]      Batch number: AE 83 CE E4 A5 
[+]   Production date: week db / 20f1

[=] --- Hardware Information
[=]    raw: 04010100011805
[=]      Vendor Id: NXP Semiconductors Germany
[=]           Type: 0x01
[=]        Subtype: 0x01
[=]        Version: 0.1 ( DESFire MF3ICD40 )
[=]   Storage size: 0x18 ( 4096 bytes )
[=]       Protocol: 0x05 ( ISO 14443-2, 14443-3 )

[=] --- Software Information
[=]    raw: 90AF0401010001
[=]      Vendor Id: no tag-info available
[=]           Type: 0xAF
[=]        Subtype: 0x04
[=]        Version: 1.1
[=]   Storage size: 0x00 ( 1 bytes )
[=]       Protocol: 0x01 ( Unknown )

[=] --------------------------------- Card capabilities ---------------------------------
[#] switch_off
[usb] pm3 --> hf mfdes auth -n 0 -t 3tdea -k 000000000000000000000000000000000000000000000000 -v -c native -a
[=] Key num: 0 Key algo: 3tdea Key[24]: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[=] Secure channel: n/a Command set: native Communication mode: plain
[+] Setting ISODEP -> inactive
[+] Setting ISODEP -> NFC-A
[=] AID 000000 is selected
[=] Auth: cmd: 0x1a keynum: 0x00
[+] raw>> 1A 00 
[+] raw<< AF EE 91 30 1E E8 F5 84 D6 C7 85 1D 05 65 13 90 A6 C6 D5 
[#] encRndB: EE 91 30 1E E8 F5 84 D6 
[#] RndB: CA FE BA BE 00 11 22 33 
[#] rotRndB: FE BA BE 00 11 22 33 CA FE BA BE 00 11 22 33 CA 
[#] Both   : 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 FE BA BE 00 11 22 33 CA FE BA BE 00 11 22 33 CA 
[+] raw>> AF 30 EB 55 F3 29 39 04 96 77 88 CE EF 33 A3 C8 7B 18 66 1A F1 62 78 A0 28 53 84 67 98 7C BB DB 03 
[+] raw<< 00 9B 71 57 8F FB DF 80 A8 F6 EF 33 4A C6 CD F9 7A 7D BE 
[=] Session key : 01 02 03 04 CA FE BA BE 07 08 09 10 22 33 CA FE 13 14 15 16 00 11 22 33 
[=] Desfire  authenticated
[+] PICC selected and authenticated succesfully
[+] Context: 
[=] Key num: 0 Key algo: 3tdea Key[24]: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 
[=] Secure channel: ev1 Command set: native Communication mode: plain
[=] Session key [24]: 01 02 03 04 CA FE BA BE 07 08 09 10 22 33 CA FE 13 14 15 16 00 11 22 33  
[=]     IV [8]: 00 00 00 00 00 00 00 00 
[+] Setting ISODEP -> inactive
```

**Comment by colinoflynn** [comment]
With both `make desfire` and `make desfire-dev`  I get this error:

```
BuildScripts/lufa_build_extra.mk:131: The XMEGA device support is currently EXPERIMENTAL (incomplete and/or non-functional), and is included for preview purposes only.
basename: extra operand ‘../LUFA/Drivers/USB/Core/DeviceStandardReq.c’
Try 'basename --help' for more information.
make: *** No rule to make target 'Bin/HIDParser.o', needed by 'Chameleon-Mini.elf'.  Stop.
```

There's no line number so not sure (it looks like maybe from an included makefile?), do you know offhand where things were tweaked? I'm building it on Ubuntu (using WSL).

**Comment by maxieds** [comment]
@colinoflynn 
Please check one more time with the latest commit. If that doesn't work, you can try testing with the binaries I just compiled in this archive: [latest-desfire-build.zip](https://github.com/emsec/ChameleonMini/files/8857676/latest-desfire-build.zip). 

BTW, thanks for the proxmark offer. My advisor gave me some money from her grant this Spring to buy one :)

**Comment by colinoflynn** [comment]
Got it! Will test those binaries - the output now fixes the basename problem, but I still get:

```
make: *** No rule to make target 'Bin/HIDParser.o', needed by 'Chameleon-Mini.elf'.  Stop.
```

How are you building that doesn't trigger the errors?

**Comment by maxieds** [comment]
@colinoflynn 
Hmmm. Did you try running ``make clean`` first? If not, I will try to debug with my Arch Linux box later.

**Comment by maxieds** [comment]
Nevermind. It should work now. 

It looks like ``brew`` distinguishes the standard system ``make`` with its version called ``gmake`` on MacOS. The Linux ``make`` command is (looks like) equivalent to ``gmake`` on Mac. Now we know...

**Comment by maxieds** [comment]
The ``make`` build problems should now all be resolved. As a bonus, running ``make desfire`` or ``make desfire-dev`` should be quicker on the average: I convinced the ``Makefile`` to not rebuild the LUFA library objects at each fresh compile of the Chameleon sources. Some might also checkout the predefined custom build targets in [this build script](https://github.com/emsec/ChameleonMini/blob/39de8cc51c4d5b949166749fac9ec839a9ac4967/Firmware/Chameleon-Mini/BuildScripts/custom_build_targets.mk#L1) to start on the idea from #283. 

@colinoflynn @lvandenb 
I have some technical problems with the code in the latest commit interfacing with the USB/PCSC ACS-122U reader. The PM3 compatibility still works, but the USB reader will not recognize the Chameleon with ``CONFIG=MF_DESFIRE`` like it did several months ago. I can trace the problem down to where the reader/PCD issues the Chameleon a DESELECT (``0x2E``) command:

What response should it actuay be sending back? This happens after the anticollision loop and RATS are satisfied. My start to fixing this issue is [found here](https://github.com/emsec/ChameleonMini/blob/39de8cc51c4d5b949166749fac9ec839a9ac4967/Firmware/Chameleon-Mini/Application/DESFire/DESFireISO14443Support.c#L436) -- all commented out again after failed tests of the code.

**Comment by maxieds** [comment]
Here are current binary builds: [latest-desfire-build.zip](https://github.com/emsec/ChameleonMini/files/8916328/latest-desfire-build.zip)

**Comment by colinoflynn** [comment]
I'm travelling for some events - I did throw  PM3/CM in my bag in case I've got time, but it's likely I'll be away from this for a couple weeks, will update if I get a chance to try. Unfortunately no other reader with me so on the final problem I can take a look once I've back to office.

**Comment by maxieds** [comment]
If anyone else wants to help out with testing PR #319, it would be nice to add some example DESFire configuration dumps generated with non-default interactions with the Chameleon. Also, posting this request as a follow up on the PR thread.

**Comment by maxieds** [comment]
@david-oswald @fptrs 
I think we can close this issue now that PR #319 is merged. I will check the latest DESFire firmware build with my two local PCSC external USB readers later tonight. If there are still problems getting ``pcsc_scan`` to recognize the Chameleon running ``CONFIG=MF_DESFIRE``, I will start a new issue where we can follow up. 😸

**Comment by colinoflynn** [comment]
Thanks for more updates! I'll still try and test them, my trip got messed up due to COVID/travel complications so I still didn't get around to it... but if I find any problems will work through them on new issue.
